task/task.go

```diff
@@ -0,0 +1,117 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package task
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// NewPreemptiveTaskManager provides a task manager which manages prioritized
+// and preemptive task execution. Preemptive tasks are some tasks aren't
+// important and we don't want them to use compute resources(CPU, NW, etc...)
+// during more important tasks(=prioritized tasks) running.
+func NewPreemptiveTaskManager(concurrency int64, interval time.Duration) *PreemptiveTaskManager {
+	return &PreemptiveTaskManager{
+		preemptiveSem:           semaphore.NewWeighted(concurrency),
+		prioritizedTaskInterval: interval,
+		prioritizedTaskCond:     sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+type PreemptiveTaskManager struct {
+	prioritizedTasks        int64
+	preemptiveSem           *semaphore.Weighted
+	prioritizedTaskInterval time.Duration
+	prioritizedTaskCond     *sync.Cond
+}
+
+// DoPrioritizedTask tells the manager that we are running a prioritized task.
+func (ts *PreemptiveTaskManager) DoPrioritizedTask() {
+	atomic.AddInt64(&ts.prioritizedTasks, 1)
+	ts.prioritizedTaskCond.Broadcast()
+}
+
+// DoPrioritizedTask tells the manager that we've done a prioritized task.
+func (ts *PreemptiveTaskManager) DonePrioritizedTask() {
+	atomic.AddInt64(&ts.prioritizedTasks, -1)
+}
+
+// InvokePreemptiveTask invokes a preemptive task. The task is started only when
+// no prioritized tasks are running. Prioritized task's execution stops the
+// execution of all preemptive tasks. Priemptive task must be immutable and
+// resumable because the manager suddenly stops and restarts the task.
```

Comment:
Needs more godoc about how a task can be stopped and restarted?

Reply:
Added more description about this manager.